compiler/rustc_mir/src/transform/check_consts/ops.rs

```diff
@@ -216,14 +216,12 @@ impl NonConstOp for FnPtrCast {
 #[derive(Debug)]
 pub struct Generator;
 impl NonConstOp for Generator {
-    fn status_in_item(&self, ccx: &ConstCx<'_, '_>) -> Status {
-        // FIXME: This means generator-only MIR is only forbidden in const fn. This is for
-        // compatibility with the old code. Such MIR should be forbidden everywhere.
-        mcf_status_in_item(ccx)
+    fn status_in_item(&self, _: &ConstCx<'_, '_>) -> Status {
+        Status::Forbidden
     }
 
     fn emit_error(&self, ccx: &ConstCx<'_, '_>, span: Span) {
-        mcf_emit_error(ccx, span, "const fn generators are unstable");
+        ccx.tcx.sess.struct_span_err(span, "Generators and `async` functions cannot be `const`").emit();
```

Comment:
Since this is already checked in the syntax based checks, couldn't this also be a `delay_span_bug`?

Reply:
I couldn't see a way to make a `DiagnosticBuilder` that gets emitted as a `delay_span_bug`. I could either inline the `delay_span_bug`, since the structured error is only emitted in one place, or wait until I change to `Result<ErrorReported, DiagnosticBuilder>`.

Reply:
You invoke `delay_as_bug` instead of `emit`: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.DiagnosticBuilder.html#method.delay_as_bug